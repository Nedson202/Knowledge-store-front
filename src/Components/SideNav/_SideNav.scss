.sidenav {
  height: 100%;
  width: 270px;
  position: fixed;
  z-index: 1;
  top: 0;
  left: 0;
  background-color: $white_1;
  overflow-x: hidden;
  padding-top: 50px;
  z-index: 10;
  box-shadow: 0 1px 1px 100vw rgba(0, 0, 0, 0.6);

  div {
    padding: 8px 8px 8px 5px;
    text-decoration: none;
    font-size: 20px;
    display: block;
    font-weight: 100;
  }

  &-auth-buttons {
    padding-left: 20px !important;
    padding-bottom: 20px !important;
    border-bottom: 1px solid $blue_2;
    margin-bottom: 15px;
  }

  &-user-details {
    padding-bottom: 20px;
    border-bottom: 1px solid $blue_2;
    margin-bottom: 15px;

    .avatar {
      height: 70px;
      width: 70px;
      display: block;
      margin: auto;
      border: 1px solid;
    }

    .username {
      text-transform: capitalize;
      font-size: 20px;
      text-align: center;
    }
  }
}

@media screen and (max-width: 560px) {
  .sidenav {
    padding-top: 15px;

    div {
      font-size: 18px;
    }

    a.dropdown-item.active {
      border-left: 2px solid $blue_2;
      background-color: transparent;
      color: $black_2;
    }
  }
}


.sidenav +
.navbar,

.sidenav +
.navbar +
.leftsidebar +
.container-content {
  pointer-events: none;
}
