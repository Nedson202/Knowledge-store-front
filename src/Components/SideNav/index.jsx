import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Avatar from '../ReviewCard/Avatar';
import { LOGOUT, LEFT_SIDEBAR_NAV_LINKS } from '../../settings';
import { logOutUser } from '../../redux/actions/userActions';

const SideNav = ({ isAuthenticated, user, dispatch }) => {
  const authNavLinks = () => {
    if (isAuthenticated) {
      return;
    }

    return (
      <div className="sidenav-auth-buttons">
        <button
          type="button"
          className="btn btn-default btn-raised cancel-button btn"
          data-toggle="modal"
          data-target="#LoginFormModal"
          id="login-button"
        >
          Login
        </button>
        <button
          type="button"
          className="btn btn-primary btn-raised text-case login-button"
          data-toggle="modal"
          data-target="#SignUpFormModal"
          id="signup-button"
        >
          Signup
        </button>
      </div>
    );
  };

  const userAvatar = () => {
    if (!isAuthenticated) {
      return;
    }

    const { username, picture, avatarColor, } = user;
    let avatar = <Avatar user={username} color={avatarColor} />;

    if (picture) {
      avatar = <img src={picture} alt="Avatar" className="avatar" />;
    }

    return (
      <div className="sidenav-user-details">
        <div>
          {avatar}
        </div>
        <p className="username">
          {username}
        </p>
      </div>
    );
  };

  const renderNavLinks = () => {
    const navLinksMap = LEFT_SIDEBAR_NAV_LINKS.map((navLink) => {
      const {
        key, label, link
      } = navLink;

      return (
        <div key={key}>
          <NavLink
            to={link}
            className={`
              dropdown-item sidebar-navlink
              ${key !== 0 && !isAuthenticated && 'blur'}
            `}
          >
            {label}
          </NavLink>
        </div>
      );
    });

    return navLinksMap;
  };

  const handleLogout = () => {
    window.localStorage.setItem(LOGOUT, true);
    logOutUser(dispatch);
  };

  const logoutAction = () => {
    if (!isAuthenticated) {
      return;
    }

    return (
      <div>
        <p
          onClick={handleLogout}
          className="dropdown-item sidebar-navlink"
        >
          Logout
        </p>
      </div>
    );
  };

  return (
    <div id="mySidenav" className="sidenav">
      {authNavLinks()}
      {userAvatar()}
      {renderNavLinks()}
      {logoutAction()}
    </div>
  );
};

SideNav.propTypes = {
  isAuthenticated: PropTypes.bool,
  user: PropTypes.object,
  dispatch: PropTypes.func,
};

SideNav.defaultProps = {
  isAuthenticated: false,
  user: {},
  dispatch: () => { },
};

const mapStateToProps = ({ auth }) => ({
  isAuthenticated: auth.isAuthenticated,
  user: auth.user,
});

export default connect(mapStateToProps)(SideNav);
