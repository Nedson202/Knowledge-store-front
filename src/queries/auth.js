import { gql } from 'apollo-boost';

const addUser = gql`
  mutation($username: String!, $password: String!, $email: String!) {
    addUser(username: $username, password: $password, email: $email) {
      token
    }
  }
`;

const verifyEmail = gql`
  mutation($token: String!) {
    verifyEmail(token: $token) {
      token,
      message
    }
  }
`;

const sendVerificationEmail = gql`
  query($email: String!) {
    sendVerificationEmail(email: $email) {
      message
    }
  }
`;

const forgotPassword = gql`
  query($email: String!) {
    forgotPassword(email: $email) {
      message
    }
  }
`;

const resetPassword = gql`
  mutation($id: String!, $email: String!, $password: String!, $token: String!) {
    resetPassword(id: $id, email: $email, password: $password, token: $token) {
      message
      token
    }
  }
`;

const editProfile = gql`
  mutation($username: String, $email: String, $picture: String) {
    editProfile(username: $username, email: $email, picture: $picture) {
      token
      message
    }
  }
`;

const changePassword = gql`
  mutation($oldPassword: String!, $newPassword: String!) {
    changePassword(oldPassword: $oldPassword, newPassword: $newPassword) {
      token
      message
    }
  }
`;

const loginUser = gql`
mutation($username: String!, $password: String!) {
  loginUser(username: $username, password: $password) {
    username,
    email,
    token
  }
}
`;

const getBooks = gql`
  {
    books {
      name,
      authors,
      genre,
      year
    }
  }
`;

export {
  addUser, loginUser, getBooks, changePassword,
  verifyEmail, sendVerificationEmail,
  forgotPassword, resetPassword, editProfile
};