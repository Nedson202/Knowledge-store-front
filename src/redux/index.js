import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';

import rootReducer from './reducers/rootReducer';

const enhancer = composeWithDevTools(applyMiddleware(thunk));

const store = createStore(
  rootReducer,
  enhancer
);

export default store;
