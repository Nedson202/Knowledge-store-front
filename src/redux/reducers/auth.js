import isEmpty from 'lodash/isEmpty';
import { SET_LOGGEDIN_USER } from '../constants/actionTypes';

const initialState = {
  isAuthenticated: false,
  user: {}
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_LOGGEDIN_USER: {
      return {
        ...state,
        isAuthenticated: action.user.id && !isEmpty(action.user),
        user: action.user
      };
    }

    default: return state;
  }
};