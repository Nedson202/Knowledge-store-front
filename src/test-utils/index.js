import React from 'react';
import { MockedProvider } from 'react-apollo/test-utils';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import store from '../redux';

// eslint-disable-next-line react/prop-types
export const AllProviders = ({ children, customStore }) => (
  <MockedProvider removeTypename>
    <Provider store={customStore || store}>
      <BrowserRouter>
        {children}
      </BrowserRouter>
    </Provider>
  </MockedProvider>
);

export * from '@testing-library/react';
