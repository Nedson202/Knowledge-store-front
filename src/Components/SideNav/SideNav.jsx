import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import './_SideNav.scss';

class SideNav extends Component {
  render() {
    return (
      <div id="mySidenav" className="sidenav">
        <div>
          <NavLink to="/my-books" className="dropdown-item sidebar-navlink">
            My books
          </NavLink>
        </div>
        <div>
          <NavLink to="/books" className="dropdown-item sidebar-navlink">
            All books
          </NavLink>
        </div>
        <div>
          <NavLink to="/profile" className="dropdown-item sidebar-navlink">
            Profile
          </NavLink>
        </div>
        <div>
          <NavLink to="/logout" className="dropdown-item sidebar-navlink">
            Logout
          </NavLink>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  user: state.auth.user
});

export default connect(mapStateToProps)(SideNav);
