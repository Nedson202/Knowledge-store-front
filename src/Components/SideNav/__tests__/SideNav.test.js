import React from 'react';
import { createStore } from 'redux';

import { render, AllProviders } from 'test-utils';
import { LEFT_SIDEBAR_NAV_LINKS } from 'settings';

import auth from '../../../redux/reducers/auth';
import SideNav from '..';

describe('SideNav', () => {
  it('should display nav links', () => {
    const { container, asFragment, getByText } = render(
      <AllProviders>
        <SideNav />
      </AllProviders>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(container).toBeInTheDocument();

    LEFT_SIDEBAR_NAV_LINKS.forEach(({ label }) => {
      getByText(label);
    });
  });

  it('should display auth buttons if not signed in', () => {
    const { getByText, queryByTestId } = render(
      <AllProviders>
        <SideNav />
      </AllProviders>
    );

    getByText('Login');
    getByText('Signup');

    expect(queryByTestId('Logout')).toBeNull();
  });

  it('should display user avatar and trigger logout option', () => {
    const mockProps = {
      auth: {
        isAuthenticated: true,
        user: {
          username: 'Cooper AL',
          picture: 'https://image.url',
          avatar: 'green',
        }
      }
    };

    const store = createStore(auth, mockProps);

    const { getByText } = render(
      <AllProviders customStore={store}>
        <SideNav />
      </AllProviders>
    );
    const logoutButton = getByText('Logout');

    getByText('Cooper AL');

    expect(logoutButton).toBeVisible();
    // fireEvent.click(logoutButton);
  });
});
