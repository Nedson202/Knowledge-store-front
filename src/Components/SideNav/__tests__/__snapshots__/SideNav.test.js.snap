// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav should display nav links 1`] = `
<DocumentFragment>
  <div
    class="sidenav"
    id="mySidenav"
  >
    <div
      class="sidenav-auth-buttons"
    >
      <button
        class="btn btn-default btn-raised cancel-button btn"
        data-target="#LoginFormModal"
        data-toggle="modal"
        id="login-button"
        type="button"
      >
        Login
      </button>
      <button
        class="btn btn-primary btn-raised text-case login-button"
        data-target="#SignUpFormModal"
        data-toggle="modal"
        id="signup-button"
        type="button"
      >
        Signup
      </button>
    </div>
    <div>
      <a
        class="
              dropdown-item sidebar-navlink
              false
            "
        href="/books"
      >
        All Books
      </a>
    </div>
    <div>
      <a
        class="
              dropdown-item sidebar-navlink
              blur
            "
        href="/my-books"
      >
        My Books
      </a>
    </div>
    <div>
      <a
        class="
              dropdown-item sidebar-navlink
              blur
            "
        href="/my-favorites"
      >
        My Favorites
      </a>
    </div>
    <div>
      <a
        class="
              dropdown-item sidebar-navlink
              blur
            "
        href="/my-profile"
      >
        My Profile
      </a>
    </div>
  </div>
</DocumentFragment>
`;
